### PR TITLE
CLI: allow setting options for config without profiles

### LINKED ISSUE #5543: verdi config should work on configurations without profiles
### Describe the bug

On an "empty" configuration (that was created automatically by AiiDA)
```json
{
    "CONFIG_VERSION": {
        "CURRENT": 9,
        "OLDEST_COMPATIBLE": 9
    },
    "profiles": {}
}
```

I'm getting the following

```
$ verdi config set warnings.development_version False --global
Traceback (most recent call last):
...
  File "/.../aiida-core/aiida/cmdline/commands/cmd_config.py", line 126, in verdi_config_set
    config: Config = ctx.obj.config
AttributeError: 'NoneType' object has no attribute 'config'
```

### Steps to reproduce

1. Take any AiiDA 2 installation
2. Point the `AIIDA_PATH` variable to an empty directory
3. Run `verdi config set warnings.development_version False`

### Expected behavior

Setting a global config option should be possible, even on a config that does not yet contain any profile.

### Your environment

- Operating system [e.g. Linux]: ubuntu
- Python version [e.g. 3.7.1]: 3.9
- aiida-core version [e.g. 1.2.1]: 2.0.1.post0

### PR DESCRIPTION
**Pull request by sphuber**
Fixes #5543 

The `verdi config set` command would except if the config had no
configured profiles. The reason is that the command would attempt to
retrieve the `Config` instance from the `ctx.obj` object. The problem is
however that this is created and set on the context in the `convert`
method of the `ProfileParamType`, which is used for the `--profile`
option.

The option specifies a default, which is the default profile, but if
that is not defined, it doesn't provide anything and so the `convert`
method is never called, resulting in the `ctx.obj` never being
initialized and the `config` not being set.

Since the config should always be present for all `verdi` comments, we
should ensure that it is always loaded and set on `ctx.obj`. Therefore
it is removed from `ProfileParamType.convert` and added to the
constructor of a custom implementation of the `click.Context` class,
called `VerdiContext`. The `VerdiCommandGroup` is updated to set the
`context_class` attribute to this `VerdiContext` which will ensure that
all commands will be invoked using a context that has the user defined
object constructure with the config loaded and added to it. Since all
commands in `verdi` will implement this class, this should guarantee
that the config will always be initialized.

### CODE REVIEW

File: aiida/cmdline/params/types/profile.py [hunk 1]
```diff
@@ -16,7 +16,12 @@
 
 
 class ProfileParamType(LabelStringType):
-    """The profile parameter type for click."""
+    """The profile parameter type for click.
+
+    This parameter type requires the command that uses it to define the ``context_class`` class attribute to be the
+    :class:`aiida.cmdline.groups.verdi.VerdiContext` class, as that is responsible for creating the user defined object
+    ``obj`` on the context and loads the instance config.
```
Comment (ltalirz) [inline aiida/cmdline/params/types/profile.py:23]: Is there a way to validate this?
If this is straightforward (even if just indirect by checking the `ctx.obj`) a check + exception might be a more effective way of enforcing this than via documentation.

File: tests/cmdline/commands/test_config.py [hunk 2]
```diff
@@ -7,143 +7,162 @@
...
+"""Tests for ``verdi config``."""
 from aiida import get_profile
 from aiida.cmdline.commands import cmd_verdi
-from aiida.manage.configuration import get_config
 
 
-class TestVerdiConfig:
-    """Tests for `verdi config`."""
+def test_config_set_option_no_profile(run_cli_command, empty_config):
+    """Test the `verdi config set` command when no ."""
```
Comment (ltalirz) [inline tests/cmdline/commands/test_config.py:16]:
```suggestion
    """Test the `verdi config set` command when no profile is present in the config."""
```

File: tests/cmdline/commands/test_config.py [hunk 3]
```diff
@@ -7,143 +7,162 @@
...
-    def setup_fixture(self, config_with_profile_factory):
-        config_with_profile_factory()
+    option_name = 'daemon.timeout'
+    option_value = str(10)
 
-    def test_config_set_option(self, run_cli_command):
-        """Test the `verdi config set` command when setting an option."""
-        config = get_config()
+    options = ['config', 'set', option_name, str(option_value)]
+    run_cli_command(cmd_verdi.verdi, options)
```
Comment (ltalirz) [inline tests/cmdline/commands/test_config.py:23]: If no profile is present, this defaults to setting on the `--global` level?
This makes sense to me as a policy, just checking.

**Review by ltalirz — COMMENTED** [review]
Fantastic, thanks a lot @sphuber for fixing this!

I only have one main comment

File: tests/cmdline/commands/test_config.py (hunk 3 above)
Comment (sphuber) [reply tests/cmdline/commands/test_config.py:23]: Indeed, if there are no profiles all option settings will always operate on global scope

File: aiida/cmdline/params/types/profile.py (hunk 1 above)
Comment (sphuber) [reply aiida/cmdline/params/types/profile.py:23]: Can do, will add it.

**Comment by sphuber** [comment]
Thanks for the review @ltalirz . I have added the check

**Review by ltalirz — APPROVED** [review]
Great - looks good to me!

**Pull request merged**